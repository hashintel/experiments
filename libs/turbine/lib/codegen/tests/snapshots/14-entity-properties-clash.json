[
  {
    "$id": "http://localhost:3000/@alice/types/entity-type/country/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
    "allOf": [],
    "description": null,
    "kind": "entityType",
    "links": {},
    "properties": {
      "http://localhost:3000/@alice/types/property-type/properties/": {
        "$ref": "http://localhost:3000/@alice/types/property-type/properties/v/1"
      }
    },
    "required": [
      "http://localhost:3000/@alice/types/property-type/properties/"
    ],
    "title": "Country",
    "type": "object"
  },
  {
    "$id": "http://localhost:3000/@alice/types/property-type/properties/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
    "description": null,
    "kind": "propertyType",
    "oneOf": [
      {
        "$ref": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1"
      }
    ],
    "title": "Name"
  },
  {
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/data-type",
    "kind": "dataType",
    "$id": "https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1",
    "title": "Text",
    "description": "An ordered sequence of characters",
    "type": "string"
  }
]
