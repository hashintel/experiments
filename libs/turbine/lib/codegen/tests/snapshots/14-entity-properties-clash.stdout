localhost_3000/alice/entity/country.rs

use alloc::string::String;

use error_stack::{Report, Result, ResultExt as _};
use hashbrown::HashMap;
use serde::Serialize;
use turbine::{
    entity::Entity, url, BaseUrl, EntityLink, EntityProperties, EntityType, EntityTypeMut,
    EntityTypeRef, GenericEntityError, OptionalEntityLink, PropertyType as _, PropertyTypeMut as _,
    PropertyTypeRef as _, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};

use crate::localhost_3000::alice::property::properties::Properties as Properties0;
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Properties {
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/properties/")]
    pub properties: Properties0,
}
impl Properties {
    fn try_from_value(
        mut properties: HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        let properties = 'property: {
            let value =
                properties.remove("http://localhost:3000/@alice/types/property-type/properties/");
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/properties/",
                )));
            };
            let value =
                <Properties0>::try_from_value(value).change_context(GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/properties/",
                ));
            value
        };
        let __report0 = turbine::fold_tuple_reports((properties,));
        let ((properties,),) = turbine::fold_tuple_reports((__report0,))?;
        let this = Self { properties };
        Ok(this)
    }

    fn as_mut(&mut self) -> PropertiesMut<'_> {
        let Self { properties } = self;
        PropertiesMut {
            properties: <Properties0 as Type>::as_mut(properties),
        }
    }

    fn as_ref(&self) -> PropertiesRef<'_> {
        let Self { properties } = self;
        PropertiesRef {
            properties: <Properties0 as Type>::as_ref(properties),
        }
    }
}
///Country
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Country {
    pub properties: Properties,
}
pub type CountryV1 = Country;
impl TypeUrl for Country {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl Type for Country {
    type Mut < 'a > = CountryMut < 'a > where Self : 'a ;
    type Ref < 'a > = CountryRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> Self::Mut<'_> {
        CountryMut {
            properties: self.properties.as_mut(),
        }
    }

    fn as_ref(&self) -> Self::Ref<'_> {
        CountryRef {
            properties: self.properties.as_ref(),
        }
    }
}
impl EntityType for Country {
    type Error = GenericEntityError;

    fn try_from_entity(value: Entity) -> Option<Result<Self, Self::Error>> {
        if Self::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = Properties::try_from_value(value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl EntityProperties for Country {
    type Properties = Properties;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for Country {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct PropertiesRef<'a> {
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/properties/")]
    pub properties: <Properties0 as Type>::Ref<'a>,
}
impl<'a> PropertiesRef<'a> {
    fn try_from_value(
        properties: &'a HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        let properties = 'property: {
            let value =
                properties.get("http://localhost:3000/@alice/types/property-type/properties/");
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/properties/",
                )));
            };
            let value = <<Properties0 as Type>::Ref<'a>>::try_from_value(value).change_context(
                GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/properties/",
                ),
            );
            value
        };
        let __report0 = turbine::fold_tuple_reports((properties,));
        let ((properties,),) = turbine::fold_tuple_reports((__report0,))?;
        let this = Self { properties };
        Ok(this)
    }

    fn into_owned(self) -> Properties {
        let Self { properties } = self;
        Properties {
            properties: <<Properties0 as Type>::Ref<'_> as TypeRef>::into_owned(properties),
        }
    }
}
///Country
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CountryRef<'a> {
    pub properties: PropertiesRef<'a>,
}
pub type CountryV1Ref<'a> = CountryRef<'a>;
impl TypeUrl for CountryRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl TypeRef for CountryRef<'_> {
    type Owned = Country;

    fn into_owned(self) -> Self::Owned {
        Country {
            properties: self.properties.into_owned(),
        }
    }
}
impl<'a> EntityTypeRef<'a> for CountryRef<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesRef::try_from_value(&value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl<'a> EntityProperties for CountryRef<'a> {
    type Properties = PropertiesRef<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for CountryRef<'_> {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}
#[derive(Debug, Serialize)]
pub struct PropertiesMut<'a> {
    #[serde(rename = "http://localhost:3000/@alice/types/property-type/properties/")]
    pub properties: <Properties0 as Type>::Mut<'a>,
}
impl<'a> PropertiesMut<'a> {
    fn try_from_value(
        properties: &'a mut HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        let properties = 'property: {
            let value = unsafe {
                let value = properties
                    .get_mut("http://localhost:3000/@alice/types/property-type/properties/");
                let value = value.map(|value| value as *mut _);
                value.map(|value: *mut serde_json::Value| &mut *value)
            };
            let Some(value) = value else {
                break 'property Err(Report::new(GenericEntityError::ExpectedProperty(
                    "http://localhost:3000/@alice/types/property-type/properties/",
                )));
            };
            let value = <<Properties0 as Type>::Mut<'a>>::try_from_value(value).change_context(
                GenericEntityError::Property(
                    "http://localhost:3000/@alice/types/property-type/properties/",
                ),
            );
            value
        };
        let __report0 = turbine::fold_tuple_reports((properties,));
        let ((properties,),) = turbine::fold_tuple_reports((__report0,))?;
        let this = Self { properties };
        Ok(this)
    }

    fn into_owned(self) -> Properties {
        let Self { properties } = self;
        Properties {
            properties: <<Properties0 as Type>::Mut<'_> as TypeMut>::into_owned(properties),
        }
    }
}
///Country
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CountryMut<'a> {
    pub properties: PropertiesMut<'a>,
}
pub type CountryV1Mut<'a> = CountryMut<'a>;
impl TypeUrl for CountryMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/country/" / v / 1u32);
}
impl TypeMut for CountryMut<'_> {
    type Owned = Country;

    fn into_owned(self) -> Self::Owned {
        Country {
            properties: self.properties.into_owned(),
        }
    }
}
impl<'a> EntityTypeMut<'a> for CountryMut<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a mut Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesMut::try_from_value(&mut value.properties.0);
        match turbine::fold_tuple_reports((properties,)) {
            Err(error) => Some(Err(error)),
            Ok((properties,)) => Some(Ok(Self { properties })),
        }
    }
}
impl<'a> EntityProperties for CountryMut<'a> {
    type Properties = PropertiesMut<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl OptionalEntityLink for CountryMut<'_> {
    fn link_data_opt(&self) -> Option<&turbine::entity::LinkData> {
        None
    }
}


---

localhost_3000/alice/property/properties.rs

use error_stack::{Report, Result, ResultExt as _};
use serde::Serialize;
use turbine::{
    types::data::Text, url, DataType, DataTypeMut, DataTypeRef, GenericPropertyError, PropertyType,
    PropertyTypeMut, PropertyTypeRef, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};
///Name
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct Properties(pub Text);
impl TypeUrl for Properties {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/properties/" / v / 1u32);
}
impl Type for Properties {
    type Mut < 'a > = PropertiesMut < 'a > where Self : 'a ;
    type Ref < 'a > = PropertiesRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> PropertiesMut<'_> {
        let Self(value) = self;
        PropertiesMut(<Text as Type>::as_mut(value))
    }

    fn as_ref(&self) -> PropertiesRef<'_> {
        let Self(value) = self;
        PropertiesRef(<Text as Type>::as_ref(value))
    }
}
impl PropertyType for Properties {
    type Error = GenericPropertyError;

    fn try_from_value(value: serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <Text as DataType>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type PropertiesV1 = Properties;
///Name
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub struct PropertiesRef<'a>(pub <Text as Type>::Ref<'a>);
impl TypeUrl for PropertiesRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/properties/" / v / 1u32);
}
impl TypeRef for PropertiesRef<'_> {
    type Owned = Properties;

    fn into_owned(self) -> Properties {
        let Self(value) = self;
        Properties(<<Text as Type>::Ref<'_> as TypeRef>::into_owned(value))
    }
}
impl<'a> PropertyTypeRef<'a> for PropertiesRef<'a> {
    type Error = GenericPropertyError;

    fn try_from_value(value: &'a serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <<Text as Type>::Ref<'a> as DataTypeRef<'a>>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type PropertiesV1Ref<'a> = PropertiesRef<'a>;
///Name
#[derive(Debug, Serialize)]
pub struct PropertiesMut<'a>(pub <Text as Type>::Mut<'a>);
impl TypeUrl for PropertiesMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/property-type/properties/" / v / 1u32);
}
impl TypeMut for PropertiesMut<'_> {
    type Owned = Properties;

    fn into_owned(self) -> Properties {
        let Self(value) = self;
        Properties(<<Text as Type>::Mut<'_> as TypeMut>::into_owned(value))
    }
}
impl<'a> PropertyTypeMut<'a> for PropertiesMut<'a> {
    type Error = GenericPropertyError;

    fn try_from_value(value: &'a mut serde_json::Value) -> Result<Self, Self::Error> {
        {
            let value = <<Text as Type>::Mut<'a> as DataTypeMut<'a>>::try_from_value(value)
                .change_context(GenericPropertyError::Data);
            value.map(Self)
        }
    }
}
pub type PropertiesV1Mut<'a> = PropertiesMut<'a>;
